tools/mason/MasonUpdate.chpl

```diff
@@ -0,0 +1,205 @@
+  }
+
+  if root.pathExists("dependencies") {
+    var deps = getDependencies(root);
+    var manifests = getManifests(deps);
+    var fullTree = createDepTrees(depTree, manifests, "root");
+    return fullTree;
+  }
+  else {
+    return depTree;
+  }
+} 
+
+proc createDepTrees(depTree: Toml, deps: [?d] Toml, name: string) : Toml { 
+  var depList: [1..0] Toml;
+  while deps.domain.size > 0 {
+    var dep = deps[deps.domain.first];
+    var brick = dep["brick"];
+    var package = brick["name"].s;
+    var version = brick["version"].s;
+    if depTree.pathExists(package) {
+      var verToUse = IVRS(version, depTree[package]["version"].s);
+      version = verToUse;
+    }
+    var source = brick["source"].s;
+    var toAdd: Toml = package + ' ' + version + ' ' + source;
+    depList.push_back(toAdd);
+
+
+    var dt: domain(string);
+    var depTbl: [dt] Toml;
+    depTree[package] = depTbl;
+    depTree[package]["name"] = package;
+    depTree[package]["version"] = version;
+    depTree[package]["source"] = source;      
+    
+    
+    if dep.pathExists("dependencies") {
+      var subDeps = getDependencies(dep);
+      var manifests = getManifests(subDeps);
+      var dependency = createDepTrees(depTree, manifests, package);
+    }
+    deps.remove(deps.domain.first);
+  }
+  depTree[name]["dependencies"] = depList;
+  return depTree;
+}
+
+
+
+/* The Incompatible Version Resolution Stategy 
+   - differing major versions are not allowed
+   - Always newest minor and patch 
+   - in accordance with semver  */
+proc IVRS(version1: string, version2: string) {
+  if version1 == version2 then return version1;
+  var vers1 = version1.split('.');
+  var vers2 = version2.split('.');
+  var v1 = vers1(1): int;
+  var v2 = vers2(1): int;
+  if vers1(1) != vers2(1) {
+    halt("Differing Major versions of dependencies are not supported");
+  }
+  else if vers1(2) != vers2(2) {
+    v1 = vers1(2): int;
+    v2 = vers2(2): int;
+    if v1 > v2 {
+      return version1;
+    }
+    else return version2;
+  }
+  else {
+    v1 = vers1(3): int;
+    v2 = vers2(3): int;
+    if v1 > v2 {
+      return version1;
+    }
+    else return version2;
+  }
+}
+
+
+
+
+/* Returns the Mason.toml for each dep listed as a Toml */
+proc getManifests(deps: [?dom] (string, Toml)) {
+  var manifests: [1..0] Toml;
+  for dep in deps {
+    var name = dep(1);
+    var version: string = dep(2).s;
+    var toAdd = retrieveDep(name, version);
+    manifests.push_back(toAdd);
+  }
+  return manifests;
+}
+
+
+/* Resposible for parsing the Mason.toml to be given
+   back to a call from getManifests */ 
+proc retrieveDep(name: string, version: string) {
+  var home = getMasonHome();
+  var tomlPath = home + "/.mason/registry/Bricks/"+name+"/"+version+".toml";
+  if isFile(tomlPath) {
+    var tomlFile = open(tomlPath, iomode.r);
+    var depToml = parseToml(tomlFile);
+    return depToml;
+  }
+  else {
+    halt("Could not find toml file for " + name);
```

Comment:
This error pops up when providing an invalid version number. I think it'd be helpful to list the version as well. If we could distinguish the errors, that'd be even better (but also could be a next step), e.g.

```
# Wrong version number
Version for PackageFoo does not exist in registry
```

```
# Wrong package name
PackageFoo does not exist in registry
```